File: DSA/src/Questions/Trie/NumberOFDistinctSubstring.java
package Questions.Trie;

public class NumberOFDistinctSubstring{
    static class Node {
        Node[] links = new Node[26]; // Trie node links

        boolean containsKey(char ch) {
            return (links[ch - 'a'] != null);
        }

        Node get(char ch) {
            return links[ch - 'a'];
        }

        void put(char ch, Node node) {
            links[ch - 'a'] = node;
        }

        // Function to insert a string into the trie
        void insert(String s) {
            Node node = this;
            // Traverse the trie, inserting characters of the string
            for (int i = 0; i < s.length(); i++) {
                char currentChar = s.charAt(i);
                // If the character is not present, create a new node
                if (!node.containsKey(currentChar)) {
                    node.put(currentChar, new Node());
                }
                // Move to the next node
                node = node.get(currentChar);
            }
        }
    }

    // Function to count the number of distinct substrings
    public static int countDistinctSubstrings(String s) {
        Node root = new Node(); // Create a trie root node
        int n = s.length();
        int count = 0;

        // Iterate through each character in the string
        for (int i = 0; i < n; i++) {
            Node node = root;

            // Iterate through each substring starting from index i
            for (int j = i; j < n; j++) {
                // If the current character is not present in the trie, insert it and increment
                // count
                if (!node.containsKey(s.charAt(j))) {
                    node.put(s.charAt(j), new Node());
                    count++;
                }
                // Move to the next node
                node = node.get(s.charAt(j));
            }
        }
        // Return the count of distinct substrings (add 1 for the empty string)
        return count + 1;
    }

    public static void main(String[] args) {
        String s1 = "ababa";
        System.out.println("Total number of distinct substrings for s1: " +
                countDistinctSubstrings(s1));

        String s2 = "ccfdf";
        System.out.println("Total number of distinct substrings for s2: " +
                countDistinctSubstrings(s2));
    }
}

File: DSA/Python/Questions/Trie/NumberOFDistinctSubstring.py
"""
# Question: Number of Distinct Substrings
# Link: https://leetcode.com/problems/number-of-distinct-substrings/

# Count distinct substrings in a string using Trie

# Time Complexity: O(n^2)
# Space Complexity: O(n^2)

# Algorithm:
# 1. Build Trie from all possible substrings
# 2. Track new nodes for distinct substrings
# 3. Count unique paths in Trie
# 4. Include empty string in final count

# Key Components:
# - TrieNode class implementation
# - count_distinct_substrings(): Main counter
# - Trie insertion and counting logic
"""


class TrieNode:
    def __init__(self):
        self.links = [None] * 26

    def contains_key(self, ch: str) -> bool:
        return self.links[ord(ch) - ord("a")] is not None

    def get(self, ch: str) -> "TrieNode":
        return self.links[ord(ch) - ord("a")]

    def put(self, ch: str, node: "TrieNode") -> None:
        self.links[ord(ch) - ord("a")] = node


def count_distinct_substrings(s: str) -> int:
    root = TrieNode()
    n = len(s)
    count = 0

    # Generate all substrings and insert into trie
    for i in range(n):
        node = root
        for j in range(i, n):
            if not node.contains_key(s[j]):
                node.put(s[j], TrieNode())
                count += 1
            node = node.get(s[j])

    # Add 1 for empty string
    return count + 1


def main():
    s1 = "ababa"
    print(
        f"Total number of distinct substrings for s1: {count_distinct_substrings(s1)}"
    )

    s2 = "ccfdf"
    print(
        f"Total number of distinct substrings for s2: {count_distinct_substrings(s2)}"
    )


if __name__ == "__main__":
    main()
